consul/consul.go

```diff
@@ -169,25 +183,53 @@ func (c *Consul) register(service *consulapi.AgentServiceRegistration) error {
 }
 
 func (c *Consul) DeregisterByTask(taskId apps.TaskId, agentAddress string) error {
-	service, err := c.findServiceByTaskId(taskId)
+	services, err := c.findServicesByTaskId(taskId)
 	if err != nil {
 		return err
+	} else if len(services) == 0 {
+		return errors.New(fmt.Sprintf("Couldn't find any service matching task id %s", taskId))
 	}
-	return c.Deregister(service)
+	return c.deregisterMultipleServices(services, taskId)
 }
 
-func (c *Consul) findServiceByTaskId(searchedTaskId apps.TaskId) (*service.Service, error) {
-	services, err := c.GetAllServices()
-	if err != nil {
-		return nil, err
-	}
+func (c *Consul) deregisterMultipleServices(services []*service.Service, taskId apps.TaskId) error {
+	var deregisterErrors []error
 	for _, s := range services {
-		taskId, err := s.TaskId()
-		if err == nil && taskId == searchedTaskId {
-			return s, nil
+		deregisterErr := c.Deregister(s)
+		if deregisterErr != nil {
+			deregisterErrors = append(deregisterErrors, deregisterErr)
 		}
 	}
-	return nil, errors.New(fmt.Sprintf("Couldn't find service matching task id %s", searchedTaskId))
+
+	return utils.MergeErrorsOrNil(deregisterErrors, fmt.Sprintf("deregistering by task %s", taskId))
+}
+
+func (c *Consul) findServicesByTaskId(searchedTaskId apps.TaskId) ([]*service.Service, error) {
+	return c.getServicesUsingProviderWithRetriesOnAgentFailure(func(agent *consulapi.Client) ([]*service.Service, error) {
+		dcAwareQueries, err := dcAwareQueriesForAllDcs(agent)
+		if err != nil {
+			return nil, err
+		}
+
+		var allFound []*service.Service
+		searchedTag := service.MarathonTaskTag(searchedTaskId)
+		for _, dcAwareQuery := range dcAwareQueries {
+			consulServices, _, err := agent.Catalog().Services(dcAwareQuery)
+			if err != nil {
```

Comment:
It would retry with a different agent. That's a general problem with the current flow also in `GetAllServices` and `GetServices` – it's either success in all DCs or nothing, with a couple of retries. We can think of a better flow for all these 3 queries.